### PR TITLE
Iss221 remove controller recursion

### LINKED ISSUE #221: Update document controller check according to spec
https://github.com/nuts-foundation/nuts-specification/pull/46

- [ ] remove recursion for finding controllers #229 
- [ ] use correct key from document #229

### PR DESCRIPTION
**Pull request by woutslakhorst**
Closes #221

There was duplication of controller resolution between VDR and ambassador. Used #227 as basis since it resolves part of the requirements.

### CODE REVIEW

File: vdr/doc/resolvers.go [hunk 1]
```diff
@@ -40,6 +40,60 @@ func (d Resolver) Resolve(id did.DID, metadata *types.ResolveMetadata) (*did.Doc
...
+			return nil, fmt.Errorf("unable to resolve controllers: %w", err)
+		}
+		leaves = append(leaves, *node)
+	}
+
+	// filter deactivated
+	j := 0
+	for i, doc := range leaves {
+		if !isDeactivated(doc) {
+			leaves[i] = doc
```
Comment (reinkrul) [inline vdr/doc/resolvers.go:85]: This way of filtering in a memory-optimized way is too smart for its own good and reeks of premature optimization.

`leaves[i] = doc` should be `leaves[j] = doc`

(see committed unit test which now fails)

File: vdr/types/interface.go [hunk 2]
```diff
@@ -35,6 +35,9 @@ type DocResolver interface {
 	// If metadata is provided then the result is filtered or scoped on that metadata.
 	// It returns ErrNotFound if there are no corresponding DID documents or when the DID Documents are disjoint with the provided ResolveMetadata
 	Resolve(id did.DID, metadata *ResolveMetadata) (*did.Document, *DocumentMetadata, error)
+	// ResolveControllers accepts a list of documents and finds their controllers
+	// The resulting list are documents who control themselves
```
Comment (reinkrul) [inline vdr/types/interface.go:39]: The addition of "The resulting list are documents who control themselves" can now be removed I think.

File: vdr/types/interface.go [hunk 3]
```diff
@@ -86,7 +89,11 @@ type KeyResolver interface {
 
 // Store is the interface that groups all low level VDR DID storage operations.
 type Store interface {
-	DocResolver
+	// Resolve returns a DID Document for the provided DID.
```
Comment (reinkrul) [inline vdr/types/interface.go:92]: "a DID Document" -> "the DID Document" (not just any, it's _the_ document)

File: vdr/vdr.go [hunk 4]
```diff
@@ -150,7 +152,7 @@ func (r VDR) Update(id did.DID, current hash.SHA256Hash, next did.Document, _ *t
 		return err
 	}
 
-	keyID := controllers[0].Authentication[0].ID.String()
+	keyID := controllers[0].CapabilityInvocation[0].ID.String()
```
Comment (reinkrul) [inline vdr/vdr.go:155]: should the alterating controller be specified in the future? Maybe we're missing a TODO/issue here.

The controller might not be controlled by the local Nuts node (edge case though).

File: vdr/vdr.go [hunk 5]
```diff
@@ -165,58 +167,5 @@ func (r VDR) Update(id did.DID, current hash.SHA256Hash, next did.Document, _ *t
...
-	if err != nil {
-		return nil, err
-	}
-	// Merge local leaves and new found leaves
-	for _, leave := range newLeaves {
-		leaves = append(leaves, leave)
-	}
-
-	return leaves, nil
+	return len(document.Controller) == 0 && len(document.CapabilityInvocation) == 0
```
Comment (reinkrul) [inline vdr/vdr.go:170]: duplicate func? (also in memoryStore)

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: vdr/ambassador.go [hunk 6]
```diff
@@ -59,6 +60,7 @@ func NewAmbassador(networkClient network.Transactions, didStore types.Store) Amb
 		networkClient: networkClient,
 		didStore:      didStore,
 		keyResolver:   doc.KeyResolver{Store: didStore},
+		docResolver:   doc.Resolver{Store: didStore},
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/ambassador.go:63]: That would add 2 extra params (not needed). The resolvers are also nothing more than a wrapper. They are not real 'services'. It must also be clear that the `store` can be passed along without creating a circular dependency.

File: vdr/ambassador.go [hunk 7]
```diff
@@ -117,7 +119,7 @@ func (n *ambassador) handleCreateDIDDocument(transaction dag.SubscriberTransacti
 	}
 
 	// Check if signingKey is one of the keys embedded in the authenticationMethod
-	didDocumentAuthKeys := proposedDIDDocument.Authentication
+	didDocumentAuthKeys := proposedDIDDocument.CapabilityInvocation
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/ambassador.go:122]: indeed...
done

File: vdr/doc/creator.go [hunk 8]
```diff
@@ -112,7 +112,7 @@ func (n Creator) Create() (*did.Document, error) {
 		ID:      didID,
 	}
 
-	doc.AddAuthenticationMethod(verificationMethod)
+	doc.AddCapabilityInvocation(verificationMethod)
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/doc/creator.go:115]: already done, latest merge on e2e tests was based upon this PR

File: vdr/doc/resolvers.go [hunk 9]
```diff
@@ -40,6 +40,60 @@ func (d Resolver) Resolve(id did.DID, metadata *types.ResolveMetadata) (*did.Doc
 	return d.Store.Resolve(id, metadata)
 }
 
+// ResolveControllers accepts a list of documents and finds their controllers
+func (d Resolver) ResolveControllers(input []did.Document) ([]did.Document, error) {
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/doc/resolvers.go:44]: Indeed will change to 1

File: vdr/doc/resolvers.go (hunk 1 above)
Comment (woutslakhorst) [reply vdr/doc/resolvers.go]: it indeed failed, but is a very common thing within Go

File: vdr/vdr.go (hunk 4 above)
Comment (woutslakhorst) [reply vdr/vdr.go:155]: the `OwnedByThisNode` helper should fix that

File: vdr/vdr.go (hunk 5 above)
Comment (woutslakhorst) [reply vdr/vdr.go:170]: Yes, dependencies are a bit of a problem here. We could remove the check from the store and only do it in the VDR?

Comment (woutslakhorst) [reply vdr/vdr.go]: moved it to a func in `store`

File: vdr/doc/resolvers.go [hunk 10]
```diff
@@ -40,6 +40,53 @@ func (d Resolver) Resolve(id did.DID, metadata *types.ResolveMetadata) (*did.Doc
 	return d.Store.Resolve(id, metadata)
 }
 
+// ResolveControllers accepts a list of documents and finds their controllers
```
Comment (reinkrul) [inline vdr/doc/resolvers.go:43]: godoc still incorrect (list of documents)

**Review by reinkrul — APPROVED** [review]

File: vdr/doc/resolvers.go (hunk 10 above)
Comment (woutslakhorst) [reply vdr/doc/resolvers.go:43]: removed it, since the doc is correct on the interface

**Pull request merged**